nowcasting_utils/models/loss.py

```diff
@@ -43,3 +51,310 @@ def get_mse_exp(self, output, target):
+    """
+    w_variance = torch.sum(torch.pow(img[:, :, :, :-1] - img[:, :, :, 1:], 2))
+    h_variance = torch.sum(torch.pow(img[:, :, :-1, :] - img[:, :, 1:, :], 2))
+    loss = tv_weight * (h_variance + w_variance)
+    return loss
+
+
+class TotalVariationLoss(nn.Module):
+    def __init__(self, tv_weight: float = 1.0):
+        super(TotalVariationLoss, self).__init__()
+        self.tv_weight = tv_weight
+
+    def forward(self, x: torch.Tensor):
+        return tv_loss(x, self.tv_weight)
+
+
+class GradientDifferenceLoss(nn.Module):
+    """
+    Gradient Difference Loss that penalizes blurry images more than MSE
+    """
+
+    def __init__(self, alpha: int = 2):
+        super(GradientDifferenceLoss, self).__init__()
+        self.alpha = alpha
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor):
+        t1 = torch.pow(
+            torch.abs(
+                torch.abs(x[:, :, :, 1:, :] - x[:, :, :, :-1, :])
+                - torch.abs(y[:, :, :, 1:, :] - y[:, :, :, :-1, :])
+            ),
+            self.alpha,
+        )
+        t2 = torch.pow(
+            torch.abs(
+                torch.abs(x[:, :, :, :, :-1] - x[:, :, :, :, 1:])
+                - torch.abs(y[:, :, :, :, :-1] - y[:, :, :, :, 1:])
+            ),
+            self.alpha,
+        )
+        # Pad out the last dim in each direction so shapes match
+        t1 = F.pad(input=t1, pad=(0, 0, 1, 0), mode='constant', value=0)
+        t2 = F.pad(input=t2, pad=(0, 1, 0, 0), mode='constant', value=0)
+        loss = t1 + t2
+        return loss.mean()
+
+
+class GridCellLoss(nn.Module):
+    """
+    Grid Cell Regularizer loss from Skillful Nowcasting, see https://arxiv.org/pdf/2104.00954.pdf
+
+    """
+
+    def __init__(self, weight_fn=None):
+        super().__init__()
+        self.weight_fn = weight_fn  # In Paper, weight_fn is max(y+1,24)
+
+    def forward(self, generated_images, targets):
+        """
+        Calculates the grid cell regularizer value, assumes generated images are the mean predictions from
+        6 calls to the generater (Monte Carlo estimation of the expectations for the latent variable)
+        Args:
+            generated_images: Mean generated images from the generator
+            targets: Ground truth future frames
+
+        Returns:
+            Grid Cell Regularizer term
+        """
+        difference = generated_images - targets
+        if self.weight_fn is not None:
+            difference *= self.weight_fn(targets)
+        difference /= targets.size(1) * targets.size(3) * targets.size(4)  # 1/HWN
+        return difference.mean()
+
+
+class NowcastingLoss(nn.Module):
+    """
+    Loss described in Skillful-Nowcasting GAN,  see https://arxiv.org/pdf/2104.00954.pdf
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x, real_flag):
+        if real_flag is True:
+            x = -x
+        return F.relu(1.0 + x).mean()
+
+
+class FocalLoss(nn.Module):
```

Comment:
worth splitting into seperate file? 
Might be worth putting the TV Loss also in a seperate file, just to keep it tidy